File: LeetcodeQuestions/leetcode_challenges/december_challange/BSTIterator.js
// class BSTIterator {
//     constructor(root) {
//         this.index = -1;
//         this.sortedNodes = [];
//         this.inOrder(root)
//     }

//     next() {
//         this.index += 1
//         if (this.index < this.sortedNodes.length) {
//             return this.sortedNodes[this.index]
//         }
//     }

//     hasNext() {
//         return this.index < this.sortedNodes.length - 1;
//     }

//     inOrder(root){
//         if (!root) return;

//         this.inOrder(root.left)
//         this.orderedNodes.push(root.val)
//         this.inOrder(root.right)
//     }
// }

var BSTIterator = function(root) {
    this.index = -1;
    this.orderedNodes = [];
    this.inOrder(root)
};

BSTIterator.prototype.next = function() {
    this.index++;
    if (this.index < this.orderedNodes.length) {
        return this.orderedNodes[this.index]
    }
};

BSTIterator.prototype.hasNext = function() {
    return this.index < this.orderedNodes.length - 1;
}

BSTIterator.prototype.inOrder = function(root) {
    if (!root) return;
    this.inOrder(root.left);
    this.orderedNodes.push(root.val);
    this.inOrder(root.right);
}

File: LeetcodeQuestions/leetcode_challenges/december_challange/BSTIterator.py
class BSTIterator:

    def __init__(self, root):
        self.index = -1
        self.sorted_nodes = []
        self.inOrder(root)

    def next(self):
        self.index += 1
        if self.index < len(self.sorted_nodes):
            return self.sorted_nodes[self.index]

    def hasNext(self):
        return self.index < len(self.sorted_nodes) - 1

    def inOrder(self, root):
        if not root:
            return
        self.inOrder(root.left)
        self.sorted_nodes.append(root.val)
        self.inOrder(root.right)